lib/src/matchers.rs

```diff
@@ -269,54 +269,67 @@ impl Matcher for IntersectionMatcher<'_> {
 /// visited.
 #[derive(PartialEq, Eq, Debug)]
 struct Dirs {
-    dirs: HashMap<RepoPath, HashSet<RepoPathComponent>>,
-    files: HashMap<RepoPath, HashSet<RepoPathComponent>>,
+    entries: HashMap<RepoPathComponent, Dirs>,
+    // is_dir/is_file aren't exclusive, both can be set to true. If entries is not empty,
+    // is_dir should be set.
+    is_dir: bool,
+    is_file: bool,
 }
 
 impl Dirs {
     fn new() -> Self {
         Dirs {
-            dirs: HashMap::new(),
-            files: HashMap::new(),
+            entries: HashMap::new(),
+            is_dir: false,
+            is_file: false,
         }
     }
 
-    fn add_dir(&mut self, dir: &RepoPath) {
-        let mut dir = dir.clone();
-        let mut maybe_child = None;
-        loop {
-            let was_present = self.dirs.contains_key(&dir);
-            let children = self.dirs.entry(dir.clone()).or_default();
-            if let Some(child) = maybe_child {
-                children.insert(child);
-            }
-            if was_present {
-                break;
+    fn add(&mut self, dir: &RepoPath) -> &mut Dirs {
+        dir.components().iter().fold(self, |sub, name| {
+            // Avoid name.clone() if entry already exists.
+            if !sub.entries.contains_key(name) {
+                sub.is_dir = true;
+                sub.entries.insert(name.clone(), Dirs::new());
             }
-            match dir.split() {
-                None => break,
-                Some((new_dir, new_child)) => {
-                    maybe_child = Some(new_child.clone());
-                    dir = new_dir;
-                }
-            };
-        }
+            sub.entries.get_mut(name).unwrap()
```

Comment:
I tried that (and some other variants), but failed to convince the borrow checker.
The first `get_mut()` borrows `entries` mutably, and it persists until `child` goes out of scope (i.e. until function returns.)

Maybe I should simply rely on `.entry(name.clone())` until it appears in the perf.

Reply:
I suppose a version with an explicit `return child;` might work, but let's leave it as is until it shows up in profiling.

Reply:
afaik, `return child` assigns wider lifetime to `child` which prevents any subsequent use of `entries`. I heard it will be addressed by new borrow checker.
https://github.com/rust-lang/polonius/